clik/external/hal_cpu/source/cpu_hal.cpp

```diff
@@ -147,10 +174,40 @@ void cpu_barrier::wait(int num_threads) {
   }
 }
 
+#if defined(HAL_CPU_WI_MODE)
 void cpu_hal::kernel_entry(exec_state *exec) {
   direct_kernel_fn kernel = (direct_kernel_fn)exec->kernel_entry;
   kernel((void *)exec->packed_args, exec);
 }
+#else
+void cpu_hal::kernel_entry(exec_state *exec) {
+  direct_kernel_fn kernel = (direct_kernel_fn)exec->kernel_entry;
+
+  // Simple split across x axis by dividing the num groups by the threads and
+  // rounding up. Some threads may get zero.
```

Comment:
What are typical use cases here? Should we take into account the possibility that exec->wg.num_groups[0] may be smaller than the number of threads?

Reply:
we do - that's why some threads get zero, in this case wg_group_x will be >= to the total threads and won't execute. It's a fairly simplistic algorithm and I think it could perhaps be improved on, but I chose this to match "host". It guarantees that no thread has more than one column different to any other.

Where this is weak is when the number in the first dimension is quite small, but I felt this was good enough for now (and in part this is intended to be a demonstration that is easy to follow).

I've tried to add a bit more to the comments.

Reply:
Thanks, if this matches what we do elsewhere that's a good enough reason for me.